### PR TITLE
[Enhancement] add support for late materialization in paquet reader

### PR DESCRIPTION
**Pull request by GavinMar**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：
Fixes #

## Problem Summary(Required) ：
When reading a parquet column chunk:
1. Read columns data with filter predicates.
2. Filter the read result and save selected row indices.
3. Read other columns, and filter page with the selected row indices returned before.  Read page header and check whether some selected rows exists in current page, if no one, skip it.
4. Merge the result data returned by step 1 and step 3.

Some performance benchmark report will be added here later.

## Perf Test：
Use following queries on tpch_100g
```sql
-- l_suppkey is not prefix key, so there is a lot parquet read.
-- Q01. for tpch-100g, there are 574 rows.
select l_orderkey, l_partkey, l_suppkey, l_linenumber from lineitem where l_suppkey = 768951;

-- Q02. for tpch-100g, there are 574 rows.
select l_orderkey, l_partkey, l_suppkey, l_linenumber, l_quantity, l_extendedprice, l_discount, l_tax, l_returnflag, l_linestatus from lineitem where l_suppkey = 768951;

-- Q03. for tpch-100g, there are 574 rows.
select * from lineitem where l_suppkey = 768951;


-- l_orderkey is prefix key, so parquet read is less
-- Q04. for tpch-100g, there are 6 rows.
select l_orderkey, l_partkey, l_suppkey, l_linenumber from lineitem where l_orderkey = 575233123;

-- Q05. for tpch-100g, there are 6 rows.
select l_orderkey, l_partkey, l_suppkey, l_linenumber, l_quantity, l_extendedprice, l_discount, l_tax, l_returnflag, l_linestatus from lineitem where l_orderkey = 575233123;

-- Q06. for tpch-100g, there are 6 rows.
select * from lineitem where l_orderkey = 575233123;

-- Q07
-- predicates on lineitem is only l_orderkey > 0, which means have to scan all rows
select * from supplier join lineitem on s_suppkey = l_suppkey where s_name = 'Supplier#000768951' and l_orderkey  > 0;
```
As the late materialization performance may be effected by coalesce read, so we test in different situations.

* parquet_coalesce_read_enable=false

Query | Disable LM | Enable LM
-- | -- | --
Q01 | 6548 | 6173
Q02 | 22574 | 5246
Q03 | 44310 | 7078
Q04 | 1051 | 1155
Q05 | 2385 | 1341
Q06 | 5157 | 1384
Q07 | 48747 | 11147

* parquet_coalesce_read_enable=true

Query | Disable LM | Enable LM
-- | -- | --
Q01 | 2597 | 2535
Q02 | 12238 | 9065
Q03 | 21311 | 17099
Q04 | 838 | 861
Q05 | 1485 | 1595
Q06 | 2913 | 2367
Q07 | 25214 | 23964

### CODE REVIEW

File: be/src/formats/parquet/group_reader.cpp [hunk 1]
```diff
@@ -44,32 +45,67 @@ Status GroupReader::get_next(vectorized::ChunkPtr* chunk, size_t* row_count) {
...
     }
 
     // other filter that not dict
     if (has_more_filter) {
         SCOPED_RAW_TIMER(&_param.stats->expr_filter_ns);
-        RETURN_IF_ERROR(ExecNode::eval_conjuncts(_left_conjunct_ctxs, _read_chunk.get()));
-        _read_chunk->check_or_die();
+        vectorized::FilterPtr conjunct_filter;
+        RETURN_IF_ERROR(ExecNode::eval_conjuncts(_left_conjunct_ctxs, active_chunk.get(), &conjunct_filter));
+        DCHECK(_merge_filter(conjunct_filter.get(), chunk_filter));
```
[the remark this answers is not in the document]
Comment (GavinMar) [reply be/src/formats/parquet/group_reader.cpp]: Sorry, I misunderstood the usage of this macro

File: be/src/formats/parquet/stored_column_reader.cpp [hunk 2]
```diff
@@ -227,11 +233,15 @@ Status RepeatedStoredColumnReader::read_records(size_t* num_records, ColumnConte
     return Status::OK();
 }
 
-Status RepeatedStoredColumnReader::_next_page() {
+// For repeated columns, it is difficult to skip rows according the num_values in header because
```
Comment (DorianZheng) [inline be/src/formats/parquet/stored_column_reader.cpp:236]: we can use def level and rep level to calculate how many rows in this page

File: be/src/formats/parquet/column_chunk_reader.cpp [hunk 3]
```diff
@@ -51,19 +52,56 @@ Status ColumnChunkReader::init(int chunk_size) {
     RETURN_IF_ERROR(get_block_compression_codec(compress_type, &_compress_codec));
 
     RETURN_IF_ERROR(_try_load_dictionary(chunk_size));
-    RETURN_IF_ERROR(_parse_page_data());
     return Status::OK();
 }
 
 Status ColumnChunkReader::next_page() {
```
Comment (DorianZheng) [inline be/src/formats/parquet/column_chunk_reader.cpp:58]: `load_header` is more accurate

File: be/src/formats/parquet/stored_column_reader.cpp (hunk 2 above)
Comment (GavinMar) [reply be/src/formats/parquet/stored_column_reader.cpp:236]: Yes, it is feasible. But we need to read and decompress page data, then parse rep level to check whether select, I'm not sure whether it is worthy.

File: be/src/formats/parquet/group_reader.cpp [hunk 4]
```diff
@@ -44,32 +45,67 @@ Status GroupReader::get_next(vectorized::ChunkPtr* chunk, size_t* row_count) {
...
+
+        size_t num_rows = lazy_chunk->num_rows();
+        lazy_chunk->filter_range(*chunk_filter, 0, num_rows);
+        if (lazy_chunk->num_rows() != *row_count) {
+            return Status::InternalError("Unmatched row count between active and lazy column chunks");
+        }
+        active_chunk->merge(std::move(*lazy_chunk));
+    }
+
+    // reorder columns as the column order of chunk is required to be invariable
```
Comment (DorianZheng) [inline be/src/formats/parquet/group_reader.cpp:99]: https://github.com/StarRocks/starrocks/pull/8597

I will reorder column in `HiveDataSource`, so you don't have to worried about it anymore

Comment (GavinMar) [reply be/src/formats/parquet/group_reader.cpp:99]: Great！I'll removing related codes in this PR.

**Comment by DorianZheng** [comment]
plz add more stats number. eg, how much page skipped, how much row group skipped...

File: be/src/formats/parquet/stored_column_reader.cpp [hunk 5]
```diff
@@ -424,14 +452,43 @@ Status OptionalStoredColumnReader::_read_records_only(size_t* num_records, Colum
...
-    } while (_num_values_left_in_cur_page == 0);
+        size_t num_values = _reader->num_values();
+        if (num_values == 0) {
+            RETURN_IF_ERROR(_reader->skip_page());
+            continue;
+        }
+        if (_select_page(num_values)) {
+            RETURN_IF_ERROR(_reader->load_page());
+            _num_values_left_in_cur_page = num_values;
+            _opts.context->next_row += num_values;
```
Comment (DorianZheng) [inline be/src/formats/parquet/stored_column_reader.cpp:467]: Do you think we need a method `ColumnReaderContext::fetch_add(int num)` for simplification? I'm afraid sometimes we forget to advance the `ColumnReaderContext::next_row`.

File: be/src/formats/parquet/stored_column_reader.cpp [hunk 6]
```diff
@@ -477,14 +543,43 @@ Status RequiredStoredColumnReader::read_records(size_t* num_records, ColumnConte
...
     return Status::OK();
 }
 
+bool RequiredStoredColumnReader::_select_page(size_t num_values) {
+    auto selection = _opts.context->selection;
+    if (!selection) {
+        return true;
+    }
+
+    size_t start_row = _opts.context->next_row;
```
Comment (DorianZheng) [inline be/src/formats/parquet/stored_column_reader.cpp:578]: Do you think we need a method `ColumnReaderContext::fetch_add(int num)` for simplification? I'm afraid sometimes we forget to advance the `ColumnReaderContext::next_row`.

Comment (GavinMar) [reply be/src/formats/parquet/stored_column_reader.cpp:578]: Yes, it seems simpler and clearer, I will modify it.

File: be/src/formats/parquet/column_chunk_reader.cpp (hunk 3 above)
Comment (dirtysalt) [reply be/src/formats/parquet/column_chunk_reader.cpp:58]: Yes. Or rename `_parse_page_header` to `load_page_header` to save a function.

File: be/src/formats/parquet/column_chunk_reader.cpp [hunk 7]
```diff
@@ -51,19 +52,56 @@ Status ColumnChunkReader::init(int chunk_size) {
     RETURN_IF_ERROR(get_block_compression_codec(compress_type, &_compress_codec));
 
     RETURN_IF_ERROR(_try_load_dictionary(chunk_size));
```
Comment (dirtysalt) [inline be/src/formats/parquet/column_chunk_reader.cpp:54]: Consider to lazy load dictionary.

File: be/src/formats/parquet/column_chunk_reader.cpp [hunk 8]
```diff
@@ -51,19 +52,56 @@ Status ColumnChunkReader::init(int chunk_size) {
...
+    uint32_t compressed_size = header.compressed_page_size;
+    uint32_t uncompressed_size = header.uncompressed_page_size;
+    size_t size = _compress_codec != nullptr ? compressed_size : uncompressed_size;
+    RETURN_IF_ERROR(_page_reader->skip_bytes(size));
+
+    _page_parse_state = PAGE_DATA_PARSED;
     return Status::OK();
 }
 
 Status ColumnChunkReader::_parse_page_header() {
```
Comment (dirtysalt) [inline be/src/formats/parquet/column_chunk_reader.cpp:90]: Yes, rename it to `load_page_header`.

File: be/src/formats/parquet/column_reader.h [hunk 9]
```diff
@@ -12,13 +12,19 @@ class HdfsScanStats;
 } // namespace starrocks
 
 namespace starrocks::parquet {
+struct ColumnReaderContext {
+    vectorized::Buffer<uint8_t>* selection = nullptr;
+    size_t next_row = 0;
+};
+
```
Comment (dirtysalt) [inline be/src/formats/parquet/column_reader.h:19]: Could you put those files into ColumnReaderOptions. No need for more struct.

Comment (dirtysalt) [reply be/src/formats/parquet/column_reader.h:19]: Another question is, how do you guarantee `selection` is released finally?

File: be/src/formats/parquet/group_reader.cpp [hunk 10]
```diff
@@ -398,4 +460,47 @@ Status GroupReader::_dict_decode(vectorized::ChunkPtr* chunk) {
     }
     return Status::OK();
 }
-} // namespace starrocks::parquet
\ No newline at end of file
+
+Status GroupReader::_merge_filter(vectorized::Filter* filter, vectorized::Filter*& dst_filter) {
```
[the remark this answers is not in the document]
Comment (dirtysalt) [reply be/src/formats/parquet/group_reader.cpp:464]: You could just simply call `&`. It's faster.

File: be/src/formats/parquet/stored_column_reader.cpp [hunk 11]
```diff
@@ -149,7 +148,8 @@ class RequiredStoredColumnReader : public StoredColumnReader {
     }
 
 private:
-    Status _next_page();
+    Status _next_selected_page(size_t records_to_read, size_t* records_to_skip);
```
Comment (dirtysalt) [inline be/src/formats/parquet/stored_column_reader.cpp:151]: What does this function mean?

File: be/src/formats/parquet/stored_column_reader.cpp [hunk 12]
```diff
@@ -477,14 +549,43 @@ Status RequiredStoredColumnReader::read_records(size_t* num_records, ColumnConte
     return Status::OK();
 }
 
-Status RequiredStoredColumnReader::_next_page() {
+Status RequiredStoredColumnReader::_next_selected_page(size_t records_to_read, size_t* records_to_skip) {
```
Comment (dirtysalt) [inline be/src/formats/parquet/stored_column_reader.cpp:552]: Is there any difference to `OptionalStoredColumnReader::_next_selected_page`?

**Comment by dirtysalt** [comment]
You could refer to this PR to design benchmark cases.

https://github.com/StarRocks/starrocks/pull/3479

File: be/src/formats/parquet/column_reader.h (hunk 9 above)
Comment (GavinMar) [reply be/src/formats/parquet/column_reader.h:21]: > Could you put those files into ColumnReaderOptions. No need for more struct.

Now we pass 'const ColumnReaderOptions&' level by level,  if the context members are put into it,  the value can not be update directly.

Comment (GavinMar) [reply be/src/formats/parquet/column_reader.h:21]: > > Could you put those files into ColumnReaderOptions. No need for more struct.
> 
> Now we pass 'const ColumnReaderOptions&' level by level, if the context members are put into it, the value can not be update directly.

Now the selection (updated to 'filter') points to a local variable, and the object will  be released automatically once the function finish.

Comment (GavinMar) [reply be/src/formats/parquet/column_reader.h:21]: > Another question is, how do you guarantee `selection` is released finally?

Now the selection (updated to 'filter') points to a local variable, and the object will be released automatically once the function finish.

File: be/src/formats/parquet/column_chunk_reader.cpp (hunk 8 above)
Comment (GavinMar) [reply be/src/formats/parquet/column_chunk_reader.cpp:95]: > Yes, rename it to `load_page_header`.

The word 'parse' can correspond to the state 'PAGE_HEADER_PARSED'.

File: be/src/formats/parquet/group_reader.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (GavinMar) [reply be/src/formats/parquet/group_reader.cpp]: Yes, this function has been removed in latest codes.

File: be/src/formats/parquet/stored_column_reader.cpp (hunk 11 above)
Comment (GavinMar) [reply be/src/formats/parquet/stored_column_reader.cpp]: > What does this function mean?

It aims to find a next valid page which satisfy the filter condition (not all rows are skipped).

File: be/src/formats/parquet/stored_column_reader.cpp (hunk 12 above)
Comment (GavinMar) [reply be/src/formats/parquet/stored_column_reader.cpp]: > Is there any difference to `OptionalStoredColumnReader::_next_selected_page`?

No difference, I have move the function to the base class.

**Review by DorianZheng — DISMISSED** [review]
LGTM

**Review by dirtysalt — DISMISSED** [review]

File: be/src/common/config.h [hunk 13]
```diff
@@ -716,6 +716,7 @@ CONF_mBool(orc_coalesce_read_enable, "true");
 CONF_mInt32(parquet_buffer_stream_reserve_size, "1048576");
 CONF_mBool(parquet_coalesce_read_enable, "true");
 CONF_mInt32(parquet_header_max_size, "16384");
+CONF_Bool(enable_parquet_late_materialization, "true");
```
Comment (dirtysalt) [inline be/src/common/config.h:719]: consider to use `parquet_late_materialization_enable`.

**Review by DorianZheng — DISMISSED** [review]
LGTM

**Review by dirtysalt — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by GavinMar** [comment]
run all

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by GavinMar** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by GavinMar** [comment]
run starrocks_admit_test

**Review by dirtysalt — APPROVED** [review]

**Review by imay — APPROVED** [review]

**Comment by dirtysalt** [comment]
run starrocks_admit_test

**Comment by GavinMar** [comment]
run starrocks_admit_test

**Pull request merged**